src/Reduce.php

```diff
@@ -0,0 +1,178 @@
+<?php
+
+declare(strict_types=1);
+
+namespace IterTools;
+
+class Reduce
+{
+    /**
+     * Reduces given collection like array_reduce() function.
+     *
+     * But unlike array_reduce(), it works with all iterable types.
+     *
+     * @template T
+     * @param iterable<mixed> $data
+     * @param callable        $reducer
+     * @param T               $initialValue
+     *
+     * @return T
+     */
+    public static function toValue(iterable $data, callable $reducer, $initialValue = null)
+    {
+        $carry = $initialValue;
+
+        foreach ($data as $datum) {
+            $carry = $reducer($carry, $datum);
+        }
+
+        return $carry;
+    }
+
+    /**
+     * Reduces given iterable to it's min value.
+     *
+     * Items of given collection must be comparable.
+     *
+     * Returns null if given collection is empty.
+     *
+     * @param iterable<mixed> $data
+     *
+     * @return mixed|null
+     */
+    public static function toMin(iterable $data)
+    {
+        return static::toValue($data, static function ($carry, $datum) {
+            return min($carry ?? $datum, $datum);
+        });
+    }
+
+    /**
+     * Reduces given iterable to it's max value.
+     *
+     * Items of given collection must be comparable.
+     *
+     * Returns null if given collection is empty.
+     *
+     * @param iterable<mixed> $data
+     *
+     * @return mixed|null
+     */
+    public static function toMax(iterable $data)
+    {
+        return static::toValue($data, static function ($carry, $datum) {
+            return max($carry ?? $datum, $datum);
+        });
+    }
+
+    /**
+     * Reduces given iterable to it's length.
+     *
+     * @param iterable<mixed> $data
+     *
+     * @return int
+     */
+    public static function toCount(iterable $data): int
+    {
+        if (is_array($data) || $data instanceof \Countable) {
```

Comment:
FYI: I think this can be replaced with a single call to [`is_countable`](https://www.php.net/manual/en/function.is-countable.php).

Reply:
Done.